posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,698 @@
+import json
+import os
+from collections.abc import Iterator
+from dataclasses import asdict, dataclass, field
+from datetime import datetime, timedelta, timezone
+from typing import Iterable, NamedTuple
+from uuid import UUID
+
+import psycopg2
+from django.conf import settings
+from temporalio import activity, workflow
+from temporalio.common import RetryPolicy
+
+from posthog.clickhouse.client.execute import sync_execute
+from posthog.temporal.workflows.base import CommandableWorkflow
+
+EPOCH = datetime(1970, 1, 1, 0, 0, tzinfo=timezone.utc)
+
+SELECT_PERSONS_TO_DELETE_QUERY = """
+SELECT
+    team_id,
+    old_person_id,
+    override_person_id,
+    max(created_at) AS latest_created_at,
+    max(version) AS latest_version,
+    min(oldest_event) AS oldest_event_at
+FROM
+    {database}.person_overrides
+WHERE
+    created_at <= %(latest_created_at)s
+GROUP BY
+    team_id, old_person_id, override_person_id
+"""
+
+SELECT_LATEST_CREATED_AT_QUERY = """
+SELECT
+    max(created_at)
+FROM {database}.person_overrides;
+"""
+
+CREATE_DICTIONARY_TABLE_QUERY = """
+CREATE OR REPLACE TABLE {database}.{dictionary_name}_table (
+    team_id INT NOT NULL,
+    old_person_id UUID NOT NULL,
+    override_person_id UUID NOT NULL
+)
+ENGINE = TinyLog();
+"""
+
+INSERT_INTO_DICTIONARY_TABLE_QUERY = """
+INSERT INTO {database}.{dictionary_name}_table
+    SELECT
+        team_id,
+        old_person_id,
+        argMax(override_person_id, created_at) AS override_person_id
+    FROM {database}.person_overrides
+    WHERE
+        created_at <= %(latest_created_at)s
+        {team_id_filter}
+    GROUP BY team_id, old_person_id;
+"""
+
+CREATE_DICTIONARY_QUERY = """
+CREATE OR REPLACE DICTIONARY {database}.{dictionary_name} ON CLUSTER {cluster_name} (
+    `team_id` INT,
+    `old_person_id` UUID,
+    `override_person_id` UUID
+)
+PRIMARY KEY team_id, old_person_id
+SOURCE(CLICKHOUSE(USER '{user}' PASSWORD '{password}' TABLE '{dictionary_name}_table' DB '{database}'))
+LAYOUT(complex_key_hashed())
+LIFETIME(0)
+"""
+
+SQUASH_EVENTS_QUERY = """
+ALTER TABLE
+    {database}.sharded_events
+UPDATE
+    person_id = dictGet('{database}.{dictionary_name}', 'override_person_id', (toInt32(team_id), person_id))
+IN PARTITION
+    %(partition_id)s
+WHERE
+    dictHas('{database}.{dictionary_name}', (toInt32(team_id), person_id))
+    {team_id_filter}
+    AND created_at <= %(latest_created_at)s;
+"""
+
+DROP_DICTIONARY_QUERY = """
+DROP DICTIONARY {database}.{dictionary_name};
+"""
+
+DROP_DICTIONARY_TABLE_QUERY = """
+DROP TABLE {database}.{dictionary_name}_table;
+"""
+
+DELETE_SQUASHED_PERSON_OVERRIDES_QUERY = """
+ALTER TABLE
+    {database}.person_overrides
+DELETE WHERE
+    old_person_id IN %(old_person_ids)s
```

Comment:
how big can _n_ get here before this blows up? probably a few things can go wrong here:

- memory (unlikely)
- max char count in query? unsure if CH has such a thing but probably